plat/xen/x86/traps.c

```diff
@@ -75,4 +87,173 @@ void traps_lcpu_init(struct lcpu *current __unused)
+	cpu_tss[idx].ist[2] =
+		(__u64) &cpu_crit_stack[idx][sizeof(cpu_crit_stack[idx])];
+
+	tss_desc = (void *) &cpu_gdt64[idx][GDT_DESC_TSS_LO];
+	tss_desc->limit_lo	= sizeof(cpu_tss[idx]);
+	tss_desc->base_lo	= (__u64) &(cpu_tss[idx]);
+	tss_desc->base_hi	= (__u64) &(cpu_tss[idx]) >> 24;
+	tss_desc->type		= GDT_DESC_TYPE_TSS_AVAIL;
+	tss_desc->p		= 1;
+
+	__asm__ __volatile__(
+		"ltr	%0\n"
+		:
+		: "r"((__u16) (GDT_DESC_OFFSET(GDT_DESC_TSS_LO))));
+}
+
+
+static struct seg_gate_desc64 cpu_idt[IDT_NUM_ENTRIES] __align(8);
+static struct desc_table_ptr64 idtptr;
+
+static inline void idt_fillgate(unsigned int num, void *fun, unsigned int ist)
+{
+	struct seg_gate_desc64 *desc = &cpu_idt[num];
+
+	/*
+	 * All gates are interrupt gates, all handlers run with interrupts off.
+	 */
+	desc->offset_hi	= (__u64) fun >> 16;
+	desc->offset_lo	= (__u64) fun & 0xffff;
+	desc->selector	= IDT_DESC_OFFSET(IDT_DESC_CODE);
+	desc->ist	= ist;
+	desc->type	= IDT_DESC_TYPE_INTR;
+	desc->dpl	= IDT_DESC_DPL_KERNEL;
+	desc->p		= 1;
+}
+
+static void idt_init(void)
+{
+	/* Ensure that traps_table_init has been called */
+	UK_ASSERT(idtptr.limit != 0);
+
+	__asm__ __volatile__("lidt %0" :: "m" (idtptr));
+}
+
+void traps_table_init(void)
+{
+	/*
+	 * Load trap vectors. All traps run on a dedicated trap stack, except
+	 * critical and debug exceptions, which have a separate stack.
+	 *
+	 * NOTE: Nested IRQs/exceptions must not use IST as the CPU switches
+	 * to the configured stack pointer irrespective of the fact that it may
+	 * already run on the same stack. For example, if we would cause a
+	 * page fault in the debug trap and both would use the same IST entry,
+	 * the page fault handler corrupts the stack (of the debug trap) and
+	 * the system eventually crashes when returning from the page fault
+	 * handler.
+	 */
+#define FILL_TRAP_GATE(name, ist)					\
+	extern void cpu_trap_##name(void);				\
+	idt_fillgate(TRAP_##name, ASM_TRAP_SYM(name), ist)
+
+	FILL_TRAP_GATE(divide_error,		2);
+	FILL_TRAP_GATE(debug,			3); /* runs on IST3 (cpu_crit_stack) */
+
+	FILL_TRAP_GATE(overflow,		2);
+	FILL_TRAP_GATE(bounds,			2);
+	FILL_TRAP_GATE(invalid_op,		2);
+	FILL_TRAP_GATE(no_device,		2);
+
+	FILL_TRAP_GATE(coproc_seg_overrun,	2);
+	FILL_TRAP_GATE(invalid_tss,		2);
+	FILL_TRAP_GATE(no_segment,		2);
+	FILL_TRAP_GATE(stack_error,		2);
+	FILL_TRAP_GATE(gp_fault,		2);
+	FILL_TRAP_GATE(page_fault,		2);
+	FILL_TRAP_GATE(spurious_int,		2);
+
+	FILL_TRAP_GATE(coproc_error,		2);
+	FILL_TRAP_GATE(alignment_check,		2);
+	FILL_TRAP_GATE(simd_error,		2);
+
+	/* xen hypervisor callback */
+	FILL_TRAP_GATE(hypervisor_callback,	2);
+
+	idtptr.limit = sizeof(cpu_idt) - 1;
+	idtptr.base = (__u64) &cpu_idt;
+}
+
+void traps_lcpu_init(struct lcpu *this_lcpu)
+{
+	int rc;
+
+	gdt_init(this_lcpu->idx);
+	tss_init(this_lcpu->idx);
+	idt_init();
+
+	/* setup xen hypervisor callback */
+	rc = hvm_set_parameter(HVM_PARAM_CALLBACK_IRQ, (2ULL << 56) | TRAP_hypervisor_callback);
```

Comment:
this magic value (`2ULL << 56 |TRAP_hypervisor_callback`) probably wants some explanation